src/client/Microsoft.Identity.Client/ManagedIdentity/ImdsCredentialProbeManager.cs

```diff
@@ -0,0 +1,111 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Net;
+using System.Net.Http;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Identity.Client.Core;
+using Microsoft.Identity.Client.Http;
+
+namespace Microsoft.Identity.Client.ManagedIdentity
+{
+    internal class ImdsCredentialProbeManager : IProbe
+    {
+        private const string CredentialEndpoint = "http://169.254.169.254/metadata/identity/credential";
+        private const string ProbeBody = ".";
+        private const string ImdsHeader = "IMDS/";
+        private static readonly SemaphoreSlim s_lock = new(1);
+        private static ProbeResult s_cachedResult;
```

Comment:
We had a lot of problems with static region auto-discovery, which uses a similar concept. 

1. Should the HTTP call fail after some time? (it does in region ... 2 seconds)
2. Are you sure that probe failure really means endpoint not available? Could the probe fail for other reasons? 
3.  should the probe (HTTP call) be retried? How often / for how long?

Reply:
we use the MSI httpmanager logic with the same retries and error conditions. 

We cannot default to credential endpoint because there are still RPs that use IMDS like AKS etc. So we need to check if credential endpoint exist first and then fallback to IMDS